### DISCUSSION TITLE
Is it possible to change the "on_command_type" "discovery_payload" item for all devices of the same type?

### DISCUSSION THREAD

**Discussion by m-a-r-k-e** [zwave-js/zwave-js-ui#3030] [Q&A] [answered]
In my setup I have multiple Fibaro Dimmer 2 devices (_Device ID: 271-4096-258 (0x010f-0x0102-0x1000)_).
I'm using MQTT Home assistant autodiscover to detect and these devices and use it. Currently my clients are Domoticz and Home Assistant.

With Home Assistant I have the problem that when I switched the light on, the value is set to 100% instead of the last used value. I figured out the reason for it. The `light_dimmer_1` one configuration (found under the "HOME ASSISTANT" tab for the node, is  following by default:
```json
{
  "type": "light",
  "object_id": "dimmer_1",
  "discovery_payload": {
    "command_topic": "zwave/95/38/1/targetValue/set",
    "state_topic": "zwave/95/38/1/currentValue",
    "state_value_template": "{{ \"OFF\" if value_json.value == 0 else \"ON\" }}",
    "brightness_command_topic": "zwave/95/38/1/targetValue/set",
    "brightness_scale": 99,
    "brightness_state_topic": "zwave/95/38/1/currentValue",
    "brightness_value_template": "{{ value_json.value }}",
    "on_command_type": "brightness",
    "device": {
      "identifiers": [
        "zwavejs2mqtt_0xf1858849_node95"
      ],
      "manufacturer": "Fibargroup",
      "model": "Dimmer 2 (FGD212)",
      "name": "Spots bureau werkkamer",
      "sw_version": "3.5"
    },
    "name": "Spots bureau werkkamer_dimmer_1",
    "unique_id": "zwavejs2mqtt_0xf1858849_95-38-1-currentValue"
  },
  "discoveryTopic": "light/Spots_bureau_werkkamer/dimmer_1/config",
  "values": [
    "38-1-currentValue",
    "38-1-targetValue"
  ],
  "persistent": false,
  "ignoreDiscovery": false,
  "id": "light_dimmer_1"
}
``` 

The problem is the default `discovery_payload` > `on_command_type` which has the value `brightness`. So acutally Home Assistant is doing exaclty what is configured.

If I change the `on_command_type` to `last` the behaviour is what I want. In that case an ON command results in the last used brightness. I tested this by changing the value using the UI and use the "Update" and "Rediscover" option.

Looking at the source I see that this value is fixed to this `brightness`: https://github.com/zwave-js/zwave-js-ui/blob/v8.11.1/hass/configurations.ts#L80

In the documentation I read that customDevices.js can be used to make changes to the configuration. 
**My question is: can I uses this to set the `on_command_type` to `last` for all my Fibaro Dimmer 2 devices? If possible I would like to keep all the other setting the default and prevent copying this. So can I import the default configuration in the customDevices.js and only override this specific value? Or is there another way to solve my problem?**

Thanks in advance.

**Comment by robertsLando** [comment]
> If possible I would like to keep all the other setting the default and prevent copying this. So can I import the default configuration in the customDevices.js and only override this specific value? Or is there another way to solve my problem?

Actually customDevices.js is used only to discover climates/fans devices that could be difficult to discover automatically, how many Fibaro devices do you have? First suggestion is to manually edit the discovery payload and store the discovery devices (see persistent column should show yes) so on next restart those discovery payloads will be used

**Reply by m-a-r-k-e** [reply]
Thanks for the extra explanation of the customDevices.js. Your question about how many Fibaro devices I have, triggers me to check if the problem was Fibaro related. That is not the case all light dimmer devices have the same issue, so also my Qubino Flush Dimmer Plus. Which is actually logical, since in the code this is all handled as “light_dimmer”.

Based on that information I take a look in the history, since I was wondering why this value was chosen then I figured out that in the past the value “last” was already used, but is was reverted later on (https://github.com/zwave-js/zwave-js-ui/pull/1360). Especially the [comment ](https://github.com/zwave-js/zwave-js-ui/pull/1360#issuecomment-870903419) about the expected behaviour was interesting. That is exactly what I also want:
| Previous | Command | Expected behavior |
|-------------|----------------|--------------------------|
| Off          | On             | Light turns on to previous brightness |
| Off          | x %            | Light turns on to x % |
| x %         | y %            | Light turns to y % |
| Any         | Off             | Light turns off |

The interesting part is that I have Domoticz and Home Assistant both configured. When I use Domoticz the behaviour is exactly as described in the table. It always works independently of the `on_command_type` value. So it looks like Domoticz is ignoring this part of the autodiscover message. Based on the referred discussion I also figured out that setting the value to `last` did not solve my problem in Home Assistant, because in that case I also have the described problems with changing the brightness.

**Conclusion for now:** setting the value to `last` is not the solution for my problem (otherwise your workaround works, since I have in total 12 dimmers). Later this week I will spend some time to analyse the messages in more detail to see the differences between Domoticz and Home Assistant, maybe that helps to find a solution that also works for Home Assistant. I will share the results here if I find something interesting.

**Reply by robertsLando** [reply]
@m-a-r-k-e thanks for sharing your findings, let me know if you find out a solution to this, will be happy to create a PR to fix it

**Reply by m-a-r-k-e** [reply]
Did a more detailed investigation and I think the most easy solution is a small change Home Assistant autodiscover. Therefore I asked a question at the Home Assistant Community: https://community.home-assistant.io/t/mqtt-light-brightness-value-when-using-the-on-button/557645

**Reply by robertsLando** [reply]
Thanks for this @m-a-r-k-e, hope they will consider your suggestion 🙏🏼 keep me updated

**Reply by m-a-r-k-e** [reply, marked as answer]
Final update about this issue. After creating a pull request in Home Assistant with my proposed solution, the Home Assistant MQTT code owners figured out the real reason for the issue. Here the summary:

The issue is that a brightness of zero is reported when the light is switched off, this can be solved by updating the `brightness_value_template` to `{{ value_json.value if value_json.value > 0 else none }}`, so a value of zero is ignored. This solves the problem in Home Assistant, but  then the dimmer is not working with Domoticz anymore. This because Domoticz is not processing the templates correctly. For more details see the discussion: https://github.com/home-assistant/core/pull/91125#discussion_r1164376784

They also came up with the idea to filter out the zero value by default, This is now implemented and approved (see https://github.com/home-assistant/core/pull/91296), so it will be available in the next Home Assistant release. That means my issue is solved without any changes in Z-Wave JS UI.

For Domoticz developers who are reading this. I think it would be very nice if MQTT autodiscovery will also support the templating correctly in the near future, especially because this is now the prefered way of using Z-Wave with Domoticz.

**Reply by robertsLando** [reply]
@m-a-r-k-e Thanks for the follow up on this, super glad you finally found a way to fix it! 💪🏼
